test/config/normalizers.test.js

```diff
@@ -0,0 +1,388 @@
+/*
+ * Copyright Elasticsearch B.V. and other contributors where applicable.
+ * Licensed under the BSD 2-Clause License; you may not use this file except in
+ * compliance with the BSD 2-Clause License.
+ */
+
+'use strict'
+
+// Test the normalizer functions
+const test = require('tape')
+const MockLogger = require('./_mock_logger')
+
+const {
+  normalizeArrays,
+  normalizeBools,
+  normalizeBytes,
+  normalizeDurationOptions,
+  normalizeIgnoreOptions,
+  normalizeInfinity,
+  normalizeKeyValuePairs,
+  normalizeNumbers,
+  normalizeElasticsearchCaptureBodyUrls,
+  normalizeDisableMetrics,
+  normalizeSanitizeFieldNames,
+  normalizeCloudProvider,
+  normalizeCustomMetricsHistogramBoundaries,
+  normalizeTransactionSampleRate
+} = require('../../lib/config/normalizers')
+
+test('#normalizeArrays()', function (t) {
+  const opts = { arrayOpt: ['a', 'b'], stringOpt: '1,2,3', numberOpt: 2 }
+
+  normalizeArrays(opts, ['arrayOpt', 'stringOpt'])
+
+  t.deepEqual(opts, { arrayOpt: ['a', 'b'], stringOpt: ['1', '2', '3'], numberOpt: 2 })
+  t.end()
+})
+
+test('#normalizeBools()', function (t) {
+  const logger = new MockLogger()
+  const fields = ['boolTrue', 'boolFalse', 'strTrue', 'strFalse', 'badWithDefault', 'badWithoutDefault']
+  const defaults = { badWithDefault: false }
+  const opts = {
+    boolTrue: true,
+    boolFalse: false,
+    strTrue: 'true',
+    strFalse: 'false',
+    badWithDefault: 'not-a-bool',
+    badWithoutDefault: 'not-a-bool',
+    nonBoolProperty: 25
+  }
+
+  normalizeBools(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    boolTrue: true,
+    boolFalse: false,
+    strTrue: true,
+    strFalse: false,
+    badWithDefault: undefined,
+    badWithoutDefault: undefined,
+    nonBoolProperty: 25
+  })
+
+  const warnings = logger.calls
+  t.ok(warnings.length === 2, 'we got warnings for bad boolean options')
+  t.deepEqual(warnings[0].interpolation, ['not-a-bool', 'badWithDefault'])
+  t.deepEqual(warnings[1].interpolation, ['not-a-bool', 'badWithoutDefault'])
+  t.end()
+})
+
+test('#normalizeBytes()', function (t) {
+  const logger = new MockLogger()
+  const fields = ['bytes', 'kiloBytes', 'megaBytes', 'gigaBytes', 'numberBytes', 'badWithDefault', 'badWithoutDefault']
+  const defaults = { badWithDefault: '25kb' }
+  const opts = {
+    bytes: '1000b',
+    kiloBytes: '100kb',
+    megaBytes: '10mb',
+    gigaBytes: '1gb',
+    numberBytes: 12345678,
+    badWithDefault: 'not-bytes',
+    badWithoutDefault: 'not-bytes',
+    anotherProperty: 25
+  }
+
+  normalizeBytes(opts, fields, defaults, logger)
+
+  t.deepEqual(opts, {
+    bytes: 1000,
+    kiloBytes: 102400,
+    megaBytes: 10485760,
+    gigaBytes: 1073741824,
+    numberBytes: 12345678,
+    badWithDefault: NaN,
```

Comment:
As for now the function does not check for a default value. There are other normalizers that fallback to default. should we do the same here?

Reply:
I think for this refactoring change you should maintain the same behaviour and that we should have a follow-up PR or issue that fixes this behaviour so that we are following: https://github.com/elastic/apm/blob/main/specs/agents/configuration.md#invalid-configuration-values